File: examples/cpp/nn_indexing.cpp
#include <vptree.hpp>
#include <vector>
#include <cmath>

struct Point
{
    double x, y;
};

using namespace vptree;

void find_nearest_neighbors(const std::vector<Point> points, const std::vector<Point> queries)
{
    const std::size_t k = 5;

    //Define a function to compute a distance between two elements
    auto square_dist = [&](const Point& a, const Point& b) -> double {
        double x2 = (a.x - b.x) * (a.x - b.x);
        double y2 = (a.y - b.y) * (a.y - b.y);
        
        return sqrt(x2 + y2);
    };
    
    //Construction complexity: O(nlog(n))
    VPTree<Point> metric_tree(points.begin(), points.end(), square_dist);
    
    //Query parameter >= 0.0
    double epsilon = 0.0;

    for(std::size_t i = 0; i < queries.size(); ++i)
    {        
        //nn.element_ptr (const Point*): indexed element pointer corresponding to found vertex
        //nn.vertex : nearest unvisited neighbor of given vertex
        //nn.distance : distance between found nearest neighbor and given vertex
        //Complexity: O(log(n)) to O(n). See README about epsilon
        nn_t<Point> nn = metric_tree.get_nearest_unvisited_neighbor(queries[i], epsilon);
        //Process 
        //nn.element_ptr...
        //nn.vertex...
        //nn.distance...

        //Complexity: O(k.log(n)) to O(kn). See README about epsilon
        std::vector<nn_t<Point>> knn = metric_tree.get_k_nearest_unvisited_neighbors(queries[i], k, epsilon);
        //Process k subelements (ascending sort by distance then by vertex ID)
        //...
    }
}

File: examples/python/nn_indexing.py
from vptree import VPTree
from typing import List, Iterable
from math import sqrt, pow
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

def find_nearest_neighbors(points : Iterable[Point], queries : List[Point]):
    k = 5

    #Define a function to compute a distance between two elements
    square_dist = lambda a, b: sqrt(pow(a.x-b.x, 2) + pow(a.y-b.y, 2))
    
    #Construction complexity: O(nlog(n))
    metric_tree = VPTree(points, square_dist)

    #Query parameter >= 0.0
    epsilon = 0.0

    for i in range(len(queries)):
        #nn.element (Point): indexed element corresponding to found vertex
        #nn.vertex (int): nearest unvisited neighbor of given vertex
        #nn.distance (float): distance between found nearest neighbor and given vertex
        #Complexity: O(log(n)) to O(n). See README about epsilon
        nn = metric_tree.get_nearest_unvisited_neighbor(queries[i], epsilon)
        #Process
        #nn.element...
        #nn.vertex...
        #nn.distance...

        knn = metric_tree.get_k_nearest_unvisited_neighbor(queries[i], k, epsilon)
        #Process k subelements